File: sample.js
// Exposed API Keys and sensitive information - CRITICAL SECURITY ISSUES!
const API_KEY = "js-super-secret-frontend-key-12345";
var PRODUCTION_DB_PASS = "prod_db_pass_dont_show"; // Using var, less scope control

// Global variables - anti-pattern, prone to conflicts
let globalCounter = 0;
window.appConfig = {
  version: "1.0",
  // Sensitive information here!
  adminEmail: "admin@example.com"
};

// Function with convoluted logic and high cognitive complexity
function calculateDiscount(price, quantity, isVIP, promoCode) {
  // Excessive nesting and unclear logic
  if (price > 0 && quantity > 0) {
    let baseTotal = price * quantity;
    let finalAmount = baseTotal;

    if (isVIP === true) { // Redundant strict equality check
      if (baseTotal > 1000) {
        finalAmount = baseTotal * 0.8; // 20% off for VIPs over 1000
      } else if (baseTotal > 500 && baseTotal <= 1000) { // Overlapping condition
        finalAmount = baseTotal * 0.85; // 15% off
      } else {
        finalAmount = baseTotal * 0.9; // 10% off
      }
    } else { // Non-VIP logic
      if (promoCode === "SAVEBIG") {
        if (baseTotal > 200) {
          finalAmount = baseTotal - 50; // $50 off
          if (finalAmount < 0) { // Defensive programming, but nested
            finalAmount = 0;
          }
        } else {
          finalAmount = baseTotal * 0.95; // 5% off
        }
      } else if (promoCode === "FREESHIP") {
        // This promo code does nothing to the price, but makes the logic complex
        console.log("Free shipping applied (no price change).");
      } else if (promoCode === null || promoCode === undefined || promoCode === "") { // Multiple checks for empty
        // No promo code, do nothing
      } else {
        // Unknown promo code
        console.warn("Unknown promo code used: " + promoCode);
      }
    }

    // Implicit type coercion - potential bug with finalAmount
    if (finalAmount == "0") { // Using == instead of ===
      console.log("Final amount is zero.");
    }

    // Inconsistent indentation
    return finalAmount;
  } else if (price <= 0) {
    // Poor error handling, returns magic number
    console.error("Price must be positive.");
    return -1;
  }
  // Missing return path for quantity <= 0
}

// Function with potential XSS vulnerability (DOM manipulation with untrusted input)
function displayUserName(userName) {
  // Directly inserting user input into innerHTML - XSS vulnerability
  document.getElementById("userDisplay").innerHTML = "Welcome, " + userName + "!";
}

// Insecure usage of eval()
function executeDynamicCode(codeString) {
  try {
    eval(codeString); // DANGER: Executes arbitrary code
  } catch (e) {
    console.error("Error executing dynamic code:", e);
  }
}

// Callback hell / deeply nested asynchronous operations
function fetchDataAndProcess(userId) {
  fetch(`/api/users/${userId}`)
    .then(response => {
      if (!response.ok) {
        throw new Error('Network response was not ok');
      }
      return response.json();
    })
    .then(userData => {
      fetch(`/api/orders/${userData.id}`)
        .then(orderResponse => {
          if (!orderResponse.ok) {
            throw new Error('Order network response was not ok');
          }
          return orderResponse.json();
        })
        .then(orderData => {
          fetch(`/api/products/${orderData.productId}`)
            .then(productResponse => {
              if (!productResponse.ok) {
                throw new Error('Product network response was not ok');
              }
              return productResponse.json();
            })
            .then(productData => {
              console.log("All data fetched:", { userData, orderData, productData });
              // Very complex logic here...
              if (userData.isAdmin) {
                if (productData.price > 100) {
                  if (orderData.status === "completed") {
                    // Do something very specific and nested
                    console.log("Admin completed high-value order for product:", productData.name);
                  }
                }
              }
            })
            .catch(error => {
              console.error("Error fetching product data:", error);
            });
        })
        .catch(error => {
          console.error("Error fetching order data:", error);
        });
    })
    .catch(error => {
      console.error("Error fetching user data:", error);
    });
}

// Unused function
function deadCodeFunction() {
  console.log("This function is never called.");
  let unusedVar = 10;
  if (true) {
    // More unused code
  }
}

// Prototype pollution vulnerability example
Object.prototype.isPolluted = "I am polluted!";

// Main execution block (simulated browser environment)
document.addEventListener("DOMContentLoaded", () => {
  console.log("Document loaded. Running JavaScript with intentional mistakes...");

  // Example calls to demonstrate issues
  let discount1 = calculateDiscount(1000, 2, true, "NONE");
  console.log("Discount 1 (VIP, large order):", discount1);

  let discount2 = calculateDiscount(150, 1, false, "SAVEBIG");
  console.log("Discount 2 (Promo, small order):", discount2);

  let discount3 = calculateDiscount(-50, 1, false, "");
  console.log("Discount 3 (Negative price):", discount3);

  // Demonstrate XSS vulnerability
  const userInput = "<img src=x onerror=alert('XSS_Attack!');>";
  // In a real scenario, this would come from user input (e.g., URL parameter, form field)
  // document.getElementById("userDisplay") would need to exist in the HTML for this to work
  // For demonstration, let's just log it if running in Node.js, otherwise assume DOM
  if (typeof document !== 'undefined' && document.getElementById("userDisplay")) {
      displayUserName(userInput);
  } else {
      console.log(`Simulating XSS: Welcome, ${userInput}!`);
  }

  // Demonstrate eval() misuse
  executeDynamicCode("console.log('Executing code via eval! Nasty!'); alert('Eval executed!');");

  // Demonstrate global variable modification
  globalCounter++;
  console.log("Global Counter:", globalCounter);
  console.log("App Config Version:", window.appConfig.version);

  // Accessing the polluted prototype property
  let emptyObj = {};
  console.log("Is object polluted?", emptyObj.isPolluted); // Will output "I am polluted!"

  // Simulate API call and callback hell
  // fetchDataAndProcess(123); // Uncomment to simulate network calls if a server is running

  // Duplicate code block
  for (let i = 0; i < 5; i++) {
    console.log("Loop iteration A:", i);
  }

  for (let i = 0; i < 5; i++) {
    console.log("Loop iteration A:", i); // This is a duplicate!
  }

  // Missing semicolon, will cause issues in some contexts
  console.log("This line might cause issues")
  console.log("Due to missing semicolon above")

});

File: sample.py
import os
import json
import logging
import sqlite3
import requests
from datetime import datetime

# Exposed API Key - CRITICAL SECURITY ISSUE!
GLOBAL_API_KEY = "sk-live-supersecretkey1234567890abcdef"
ANOTHER_SECRET = 'my_db_password_123' # Another exposed secret

# Global mutable list - anti-pattern, leads to unexpected behavior
GLOBAL_DATA_CACHE = []

logging.basicConfig(level=logging.INFO)

def process_data(data, enable_debug=True):
    """
    Processes input data with highly convoluted logic.
    This function has high cyclomatic and cognitive complexity.
    It also demonstrates poor error handling and inconsistent naming.
    """
    # Unnecessary nested if-else and redundant checks
    if data is not None:
        if isinstance(data, dict):
            if 'items' in data and len(data['items']) > 0:
                total_value = 0
                idx = 0
                # Using while loop instead of more Pythonic for loop with enumerate
                while idx < len(data['items']):
                    item = data['items'][idx]
                    if 'price' in item and 'quantity' in item:
                        try:
                            # Bad type conversion, potential for ValueError
                            price = float(item['price'])
                            qty = int(item['quantity'])

                            if price >= 0 and qty > 0: # Redundant check
                                subtotal = price * qty
                                total_value += subtotal
                                if enable_debug == True: # Redundant comparison
                                    print("Processing item:", item.get('name', 'Unknown'))
                                    logging.debug(f"Item subtotal: {subtotal}")
                                    # Inconsistent string formatting
                                    print("Current total is " + str(total_value))
                            elif price < 0 or qty <= 0: # Another redundant check
                                print("Warning: Negative price or non-positive quantity encountered.")
                                continue # Redundant continue
                            else: # Unreachable else
                                print("Should not reach here.")
                        except Exception as e: # Catching generic Exception
                            logging.error("Error processing item: %s" % e)
                            # Silently failing without proper error propagation
                            pass
                    else:
                        if True: # Always true condition - dead code
                            print("Item missing price or quantity.")
                        if not False: # Always true condition - more dead code
                            if False: # Always false condition - even more dead code
                                pass
                            else:
                                pass
                        pass # Does nothing
                    idx += 1
                return total_value
            else:
                logging.warning("No items found in data dictionary or empty list.")
                return 0
        elif isinstance(data, list):
            # Duplicate logic, should be refactored
            sum_of_numbers = 0
            for number_str in data:
                try:
                    sum_of_numbers += int(number_str) # Potential for ValueError
                except Exception as e:
                    print("Could not convert %s to int: %s" % (number_str, e))
            return sum_of_numbers
        else:
            if 1 == 1: # Redundant comparison
                logging.info("Data is neither dict nor list. Returning -1.")
            return -1
    else:
        # Returning different type based on input - anti-pattern
        return "Invalid input: Data is None"

def check_auth(token):
    """
    Simulates an authentication check.
    Exposes an API key directly in the function.
    """
    if token == GLOBAL_API_KEY:
        print("Authentication successful.")
        return True
    else:
        # Insecure error message, reveals sensitive information
        print(f"Authentication failed. Token '{token}' is incorrect.")
        return False

def make_api_call(endpoint):
    """
    Makes an API call using the exposed global API key.
    This is a severe security vulnerability.
    """
    headers = {"Authorization": f"Bearer {GLOBAL_API_KEY}"}
    try:
        response = requests.get(f"https://api.example.com/{endpoint}", headers=headers, verify=False) # Disabling SSL verification - SECURITY RISK!
        response.raise_for_status()
        return response.json()
    except requests.exceptions.RequestException as e:
        print(f"API call failed: {e}")
        return None

def connect_to_database():
    """
    Connects to a SQLite database using hardcoded credentials.
    Insecure and not robust.
    """
    try:
        conn = sqlite3.connect('my_insecure_db.db')
        cursor = conn.cursor()
        cursor.execute(f"CREATE TABLE IF NOT EXISTS users (id INTEGER PRIMARY KEY, username TEXT, password TEXT DEFAULT '{ANOTHER_SECRET}')") # SQL Injection vulnerability if password was dynamic input
        conn.commit()
        return conn
    except Exception as e:
        print(f"Database connection error: {e}")
        return None

def execute_sql_query(conn, query):
    """
    Executes a raw SQL query directly, highly susceptible to SQL injection.
    """
    cursor = conn.cursor()
    try:
        cursor.execute(query) # SQL Injection point
        conn.commit()
        return cursor.fetchall()
    except Exception as e:
        print(f"SQL execution error: {e}")
        return None

def unused_function():
    """This function is never called - dead code."""
    x = 10
    y = 20
    return x * y

def very_complex_function(a, b, c, d, e, f):
    """
    A function designed to have extremely high cyclomatic and cognitive complexity.
    Excessive nesting, multiple return points, and complex conditions.
    """
    if a > 0:
        if b < 10:
            for i in range(c):
                if i % 2 == 0:
                    while d > 0:
                        d -= 1
                        if e == "test":
                            if f is not None and len(f) > 5:
                                print("Deeply nested logic path 1")
                                return "path1"
                            else:
                                print("Deeply nested logic path 2")
                                return "path2"
                        elif e == "prod":
                            try:
                                result = 10 / d # Potential ZeroDivisionError
                                print(f"Result: {result}")
                            except ZeroDivisionError:
                                print("Division by zero in complex function")
                                return "division_error"
                            if result > 5:
                                print("Deeply nested logic path 3")
                                return "path3"
                        else:
                            print("Deeply nested logic path 4")
                            return "path4"
                else:
                    print("Odd i in loop")
        else:
            print("b is not less than 10")
            if a % 2 != 0:
                if c > 5:
                    return "path5"
                else:
                    return "path6"
            else:
                return "path7"
    elif a == 0:
        if b == 0 and c == 0:
            if d == 0 or e == "":
                return "path8"
            else:
                return "path9"
        else:
            return "path10"
    else:
        if f is None:
            return "path11"
        else:
            return "path12"
    return "default_path" # Fallback, might indicate missing return in some paths


# Main execution block
if __name__ == "__main__":
    print("Starting Python code with intentional mistakes...")

    # Data for processing
    sample_data = {
        'items': [
            {'name': 'Laptop', 'price': 1200, 'quantity': 1},
            {'name': 'Mouse', 'price': '50.5', 'quantity': 2}, # Price as string
            {'name': 'Keyboard', 'price': 75, 'quantity': 0},  # Zero quantity
            {'name': 'Headphones', 'price': -100, 'quantity': 1}, # Negative price
            {'name': 'Monitor', 'price': 300, 'qty': 1} # Missing 'quantity' key
        ],
        'metadata': 'some_info'
    }

    list_data = ['10', 'abc', '20', '30'] # Mixed types

    # Call with various problematic inputs
    result1 = process_data(sample_data)
    print(f"\nResult 1 (dict): {result1}")

    result2 = process_data(list_data)
    print(f"Result 2 (list): {result2}")

    result3 = process_data(None)
    print(f"Result 3 (None): {result3}")

    # Authentication and API call
    is_authenticated = check_auth("wrong_key")
    if is_authenticated:
        api_response = make_api_call("users/profile")
        print(f"API Response: {api_response}")

    # Database operations
    db_conn = connect_to_database()
    if db_conn:
        user_input_username = "malicious_user"
        user_input_password = "password' OR '1'='1" # SQL Injection payload
        # Insecure SQL query construction
        query_insert = f"INSERT INTO users (username, password) VALUES ('{user_input_username}', '{user_input_password}')"
        print(f"\nExecuting insecure query: {query_insert}")
        execute_sql_query(db_conn, query_insert)

        query_select = "SELECT * FROM users WHERE username = 'admin' OR 1=1" # Another SQL Injection
        print(f"Executing insecure query: {query_select}")
        users = execute_sql_query(db_conn, query_select)
        print(f"Users found: {users}")
        db_conn.close()

    # Call the very complex function
    complex_result = very_complex_function(10, 5, 3, 2, "test", "abcdefg")
    print(f"\nComplex function result: {complex_result}")

    complex_result_2 = very_complex_function(0, 0, 0, 0, "", None)
    print(f"Complex function result 2: {complex_result_2}")

    # Using global data cache in a non-thread-safe way
    GLOBAL_DATA_CACHE.append("new item")
    print(f"Global cache: {GLOBAL_DATA_CACHE}")

    # Intentional copy-paste error
    a = 1
    b = 2
    c = 3
    # This block is duplicated
    if a > b:
        print("a is greater than b")
    else:
        print("b is greater than a")

    if a > b:
        print("a is greater than b")
    else:
        print("b is greater than a") # Duplicate code detected!